File: NinePSharp.Messages.FSharp/messages/Tgetattr.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Tgetattr =
    { Size: uint32
      Tag: uint16
      Fid: uint32
      RequestMask: uint64 }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Tgetattr
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Tgetattr span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Fid); offset <- offset + 4
            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), this.RequestMask)

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let requestMask = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8))
        { Size = size; Tag = tag; Fid = fid; RequestMask = requestMask }

File: NinePSharp/messages/Tgetattr.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Tgetattr : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Tgetattr;
    public ushort Tag { get; }
    public uint Fid { get; }
    public ulong RequestMask { get; }

    public Tgetattr(uint size, ushort tag, uint fid, ulong requestMask)
    {
        Size = size;
        Tag = tag;
        Fid = fid;
        RequestMask = requestMask;
    }

    public Tgetattr(ushort tag, uint fid, ulong requestMask)
    {
        Size = NinePConstants.HeaderSize + 4 + 8;
        Tag = tag;
        Fid = fid;
        RequestMask = requestMask;
    }

    public Tgetattr(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        Fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        RequestMask = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8));
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Tgetattr);
        int offset = NinePConstants.HeaderSize;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Fid); offset += 4;
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), RequestMask);
    }
}
